### ISSUE TITLE
OS Viewer: Hierarchies not working?

### ISSUE THREAD

**Issue by skarampatakis** [openbudgets/platform#40]
@larjohn 
In this [dataset](http://eis-openbudgets.iais.fraunhofer.de/viewer/budget-thessaloniki-revenue-2011__80bce?measure=%22amount.sum%22&groups%5B%5D=%22economicClassification.notation%22&filters%5BbudgetPhase.budgetPhase%5D%5B%5D=%22http:%2F%2Fdata.openbudgets.eu%2Fresource%2Fcodelist%2Fbudget-phase%2Fapproved%22&order=%22amount.sum%7Cdesc%22&visualizations%5B%5D=%22Treemap%22&lang=en) if you choose a hierarchy for the treemap view, and click on a rectangle it seems that nothing happens. For that specific case I know that the bottom level of the the hierarchy (economic classification) is defined as dimension value on the dataset. 

Shouldn't the Viewer initialize the graph with the higher levels? So we can eventually drilldown the dataset in the deepest detail available.

**Comment by skarampatakis** [comment]
Any update?

**Comment by larjohn** [comment]
I need some more time. Please be patient.

**Comment by pwalsh** [comment]
@larjohn let us know if this is somewhat that the OKI team can help with.

**Comment by larjohn** [comment]
@skarampatakis is this information somehow explicitly available in the RDF or it should be inferred looking at the values?

**Comment by skarampatakis** [comment]
Yes, all information is available through the definition of the Concept Scheme, each codelist item belongs to.

**Comment by larjohn** [comment]
Cool, this is not implemented yet. Give me a couple of days and it will be there.

**Comment by larjohn** [comment]
@skarampatakis can you give an example (SPARQL preferred) that connects any two of those dimensions here:

http://data.openbudgets.eu/page/dataset/budget-thessaloniki-revenue-2011/slice/248

For your convenience, [here ](http://ws307.math.auth.gr/rudolf/public/sparql?query=%0D%0APREFIX+rdfs%3A+%3Chttp%3A%2F%2Fwww.w3.org%2F2000%2F01%2Frdf-schema%23%3E%0D%0APREFIX+qb%3A+%3Chttp%3A%2F%2Fpurl.org%2Flinked-data%2Fcube%23%3E%0D%0APREFIX+skos%3A+%3Chttp%3A%2F%2Fwww.w3.org%2F2004%2F02%2Fskos%2Fcore%23%3E%0D%0APREFIX+rdf%3A+%3Chttp%3A%2F%2Fwww.w3.org%2F1999%2F02%2F22-rdf-syntax-ns%23%3E%0D%0APREFIX+xro%3A+%3Chttp%3A%2F%2Fpurl.org%2Fxro%2Fns%23%3E%0D%0A%0D%0ASELECT+DISTINCT+%3Fattribute+%28GROUP_CONCAT+%28distinct+%3Flabels%3B%0D%0A%09separator+%3D%27%7C%7C%27%29+AS+%3Flabels%29+%3Fattachment+%28SAMPLE+%28%3F_propertyType%29+AS+%3FpropertyType%29+%3FshortName+%3Fdataset+WHERE+%7B%0D%0A%09%3Fdsd+qb%3Acomponent+%3Fcomponent+.%0D%0A%09%3Fdataset+a+qb%3ADataSet+%3B%0D%0A%09%09qb%3Astructure+%3Fdsd+.%0D%0A%09%3Fcomponent+%3FcomponentProperty+%3Fattribute+.%0D%0A%09%3FcomponentProperty+rdfs%3AsubPropertyOf+qb%3AcomponentProperty+.%0D%0A%09OPTIONAL+%7B%0D%0A%09%09%3Fattribute+rdfs%3Alabel+%3Flabel+.%0D%0A%09%7D%0D%0A%09OPTIONAL+%7B%0D%0A%09%09%3Fcomponent+qb%3AcomponentAttachment+%3Fattachment+.%0D%0A%09%7D%0D%0A%09OPTIONAL+%7B%0D%0A%09%09%3Fattribute+a+%3F_propertyType+.%0D%0A%09%09FILTER+%28%3F_propertyType+in+%28qb%3ACodedProperty%2C+qb%3AMeasureProperty%2C+qb%3ADimensionProperty%29%29%0D%0A%09%7D%0D%0A%09FILTER+%28%3Fname+%3D+%27budget-thessaloniki-revenue-2011__80bce%27%29+BIND+%28REPLACE+%28str+%28%3Fattribute%29%2C+%27%5E.*%28%23%7C%2F%29%27%2C+%22%22%29+AS+%3FshortName%29+BIND+%28CONCAT+%28REPLACE+%28str+%28%3Fdataset%29%2C+%27%5E.*%28%23%7C%2F%29%27%2C+%22%22%29%2C+%27__%27%2C+SUBSTR+%28MD5+%28STR+%28%3Fdataset%29%29%2C1%2C5%29%29+AS+%3Fname%29+BIND+%28CONCAT+%28LANG+%28%3Flabel%29%2C%27+%3D%27%2C+%3Flabel%29+AS+%3Flabels%29%0D%0A%7D%0D%0AGROUP+BY+%3Fattribute+%3FshortName+%3Fattachment+%3Fdataset%0D%0A&should-sponge=&format=text%2Fhtml) is the model of the dataset:

**Comment by skarampatakis** [comment]
I am not sure that I have understand what you request. Could you please explain more? 
What I originally requested was to presesnt values as aggregations of the parents in the hierarchy, if it is available of course. For instance, in the greek case, this is applicable on the obeu-dimension:economicClassification, which represent a hierarchy. 

On the dataset you have defined only the lowest level leaf. Let say kae:6421. But this is a leaf node in the hierarchy. So it goes like kae:6 -> kae:64 -> kae:642 -> kae:6421, where the arrow means ```skos:narrower``` and ```kae:6``` is a top concept of the Concept scheme. 

So if you know what is the longest path (the level of detail) you can produce aggregated view of the datasets with a well defined level of detail, along a dimension. And you could also provide drilldowns. 

So please explain a bit better of what you want, so I can give you a more detailed answer.

**Comment by larjohn** [comment]
In order for this to work, at a Babbage-compatible API, the levels of the hierarchy should be mapped to specific dimensions. Then these dimensions are groupped into a single hierarchy.

For our example, additional columns should be artificially created in the original table (or in the pipeline) and then also created in the dsd and the dataset.

**Comment by skarampatakis** [comment]
As this seems to be application specific, couldn't you just create these extra columns on Rudolf layer?

try a query like this
```

prefix gr-dimension: <http://data.openbudgets.eu/ontology/dsd/greek-municipalities/dimension/> 
prefix obeu-budgetphase: <http://data.openbudgets.eu/resource/codelist/budget-phase/> 
prefix obeu-measure: <http://data.openbudgets.eu/ontology/dsd/measure/>
prefix obeu-dimension: <http://data.openbudgets.eu/ontology/dsd/dimension/>
prefix qb: <http://purl.org/linked-data/cube#>


select ?notation2 (group_concat(distinct ?notation;separator="||") as ?hierarchies) (SUM(?amount) AS ?sum)
where
{
?dataset  obeu-dimension:fiscalYear ?year .
?dataset  obeu-dimension:organization ?organization .
?dataset qb:slice ?slice .
?slice qb:observation ?observation .
?slice gr-dimension:economicClassification ?classification .

?classification skos:broader+ ?topConcept .
?classification skos:notation ?notation2 .


?topConcept skos:prefLabel ?label .
?topConcept skos:notation ?notation.
?observation gr-dimension:budgetPhase ?phase. 
?observation obeu-measure:amount ?amount .



VALUES ?phase {obeu-budgetphase:revised}
VALUES ?organization { <http://el.dbpedia.org/resource/Δήμος_Αθηναίων>}
VALUES ?year {<http://reference.data.gov.uk/id/year/2005>}
}
group by ?notation2
order by ?notation2
```
then you could just explode the hierarchies column and get your virtual columns.  

Or perhaps there could be a better query but I think we can and should do this at the application logic and leave data as is. 

Annotating datasets with extra dimensions would produce excess triples, which is not always efficient, since these kind of info can always be retrieved by simple SPARQL queries.

**Comment by larjohn** [comment]
This could be done, but it is a major refactoring for the existing query builder so it might be left as a top priority after the deliverable due on the end of this month.
